lib/components/user/notification-prefs-pane.js

```diff
@@ -28,85 +36,393 @@ const allowedNotificationChannels = [
+    this.state = {
+      // Holds the last known verified phone number.
+      currentNumber: initialPhoneNumber,
+
+      // The verified status of the above number.
+      currentNumberVerified: initialPhoneNumberVerified,
+
+      // If true, phone number is being edited.
+      // For new users, render component in editing state.
+      isEditing: isBlank(initialPhoneNumber),
+
+      // If true, a phone verification request has been sent and the UI is awaiting the user sending the code.
+      isPending: false,
+
+      // Holds the new phone number (+15555550123 format) entered by the user
+      // (outside of Formik because (Phone)Input does not have a standard onChange event or simple valitity test).
+      newPhoneNumber: ''
+    }
+  }
+
+  _handleEditNumber = () => {
+    this.setState({
+      isEditing: true,
+      isPending: false,
+      newPhoneNumber: ''
+    })
+  }
+
+  _handleNewPhoneNumberChange = newPhoneNumber => {
+    this.setState({
+      newPhoneNumber
+    })
+  }
+
+  _handleCancelEditNumber = () => {
+    this.setState({
+      isEditing: false,
+      isPending: false
+    })
+  }
+
+  _handlePhoneNumberKeyDown = e => {
+    if (e.keyCode === 13) {
+      // On the user pressing enter (keyCode 13) on the phone number field,
+      // prevent form submission and request the code.
+      e.preventDefault()
+      this._handleRequestCode()
+    }
+  }
+
+  _handleValidationCodeKeyDown = e => {
+    if (e.keyCode === 13) {
+      // On the user pressing enter (keyCode 13) on the validation code field,
+      // prevent form submission and send the validation code.
+      e.preventDefault()
+      this._handleSubmitCode()
+    }
+  }
+
+  _handleRequestCode = async () => {
+    // Send phone verification request with the entered values.
+    const { onRequestCode, setFieldValue, setFieldTouched } = this.props
+    const { currentNumber, currentNumberVerified, newPhoneNumber } = this.state
+
+    // If user enters the same phone number as their current verified number,
+    // just cancel editing without sending a verification SMS.
+    if (newPhoneNumber === currentNumber && currentNumberVerified) {
+      this._handleCancelEditNumber()
+    } else {
+      // Empty the input box (and disable the Send text button).
+      this.setState({
+        newPhoneNumber: ''
+      })
+
+      const { isPhoneNumberVerified, phoneNumber } = await onRequestCode(newPhoneNumber)
+
+      if (phoneNumber === newPhoneNumber && !isPhoneNumberVerified) {
+        // On success (phoneNumber is updated and verified status is false), update the phone number in component state.
+        this.setState({
+          newPhoneNumber: phoneNumber
+        })
+
+        // Set empty code field in Formik's state and prompt for code.
+        setFieldValue('validationCode', '')
+        setFieldTouched('validationCode', false)
+        this.setState({
+          isEditing: false,
+          isPending: true
+        })
```

Comment:
Since this component state update relies on application state updates, I think it might make sense to move these state variables into the application state.

Reply:
`isPending` was removed in https://github.com/opentripplanner/otp-react-redux/pull/224/commits/bc43d29931e3907bca16067f26d718075ad0e6e3.